subflow.go

```diff
@@ -0,0 +1,248 @@
+}
+
+func startSubflow(to string, c net.Conn, mpc *mpConn, clientSide bool, probeStart time.Time, tracker StatsTracker) *subflow {
+	sf := &subflow{
+		to:          to,
+		conn:        c,
+		mpc:         mpc,
+		chClose:     make(chan struct{}),
+		sendQueue:   make(chan *sendFrame),
+		pendingAcks: make(map[uint64]pendingAck),
+		emaRTT:      ema.NewDuration(longRTT, rttAlpha),
+		tracker:     tracker,
+	}
+	go sf.sendLoop()
+	if clientSide {
+		initialRTT := time.Since(probeStart)
+		tracker.UpdateRTT(initialRTT)
+		sf.emaRTT.SetDuration(initialRTT)
+		// pong immediately so the server can calculate the RTT between when it
+		// sends the leading bytes and receives the pong frame.
+		sf.ack(frameTypePong)
+	} else {
+		sf.probeStart.Store(probeStart)
+	}
+	go func() {
+		if err := sf.readLoop(); err != nil {
+			log.Debugf("read loop to %s ended: %v", sf.to, err)
+		}
+	}()
+	return sf
+}
+
+func (sf *subflow) readLoop() (err error) {
+	ch := make(chan *frame)
+	r := byteReader{Reader: sf.conn}
+	go func() {
+		defer close(ch)
+		for {
+			sz, err := ReadVarInt(r)
+			if err != nil {
+				sf.close()
+				return
+			}
+			fn, err := ReadVarInt(r)
+			if err != nil {
+				sf.close()
+				return
+			}
+			if sz == 0 {
+				sf.gotACK(fn)
+				continue
+			}
+			log.Tracef("got frame %d from %s with %d bytes", fn, sf.to, sz)
+			buf := pool.Get(int(sz))
+			n, err := io.ReadFull(r, buf)
+			if err != nil {
+				pool.Put(buf)
+				sf.close()
+				return
+			}
+			if n != int(sz) {
+				panic(fmt.Sprintf("expect to read %d bytes, read %d", sz, n))
+			}
+			sf.ack(fn)
+			ch <- &frame{fn: fn, bytes: buf}
+			sf.tracker.OnRecv(sz)
+			select {
+			case <-sf.chClose:
+				return
+			default:
+				// continue
+			}
+		}
+	}()
+	probeTimer := time.NewTimer(randomize(probeInterval))
+	for {
+		select {
+		case frame := <-ch:
+			if frame == nil {
+				return
+			}
+			sf.mpc.recvQueue.add(frame)
+			if !probeTimer.Stop() {
+				<-probeTimer.C
+			}
+			probeTimer.Reset(randomize(probeInterval))
+		case <-probeTimer.C:
+			sf.probe()
+			probeTimer.Reset(randomize(probeInterval))
+		}
+	}
+}
+
+func (sf *subflow) sendLoop() {
+	for {
+		select {
+		case <-sf.chClose:
+			return
+		case frame := <-sf.sendQueue:
+			n, err := sf.conn.Write(frame.buf)
+			if err != nil {
+				sf.close()
+				if frame.isDataFrame() {
+					sf.mpc.retransmit(frame)
+				}
+				continue
+			}
+			if n != len(frame.buf) {
+				panic(fmt.Sprintf("expect to write %d bytes on %s, written %d", len(frame.buf), sf.to, n))
+			}
+			if !frame.isDataFrame() {
+				frame.release()
+				continue
+			}
+			if frame.retransmissions == 0 {
+				sf.tracker.OnSent(frame.sz)
+			} else {
+				sf.tracker.OnRetransmit(frame.sz)
+			}
+			log.Tracef("done writing frame %d with %d bytes via %s", frame.fn, frame.sz, sf.to)
+			sf.muPendingAcks.Lock()
+			sf.pendingAcks[frame.fn] = pendingAck{int(frame.sz), time.Now()}
+			sf.muPendingAcks.Unlock()
+			frameCopy := *frame // to avoid race
```

Comment:
I'm a bit confused by this.  What would the race be?

Reply:
`time.AfterFunc` accesses the frame from another goroutine but I can't remember the exact place race happened. Will reproduce.

Reply:
No longer see data race if use `frame` directly. Maybe because of some later changes. Will remove the copy to avoid confusing.